services/dmarc-report/src/calculate-percentages.js

```diff
@@ -4,17 +4,73 @@ const calculatePercentages = ({ fail, pass, passDkimOnly, passSpfOnly }) => {
     0,
   )
 
+  // calculate base percentages
+  const failPercentage = fail <= 0 ? 0 : Number((fail / total) * 100)
+
+  const passPercentage = pass <= 0 ? 0 : Number((pass / total) * 100)
+
+  const passDkimOnlyPercentage =
+    passDkimOnly <= 0 ? 0 : Number((passDkimOnly / total) * 100)
+
+  const passSpfOnlyPercentage =
+    passSpfOnly <= 0 ? 0 : Number((passSpfOnly / total) * 100)
+
+  const percentages = [
+    failPercentage,
+    passPercentage,
+    passDkimOnlyPercentage,
+    passSpfOnlyPercentage,
+  ]
+
+  // do a basic rounding of all percentages
+  const rounded = percentages.map((x) => Math.floor(x))
+
+  // get the sum of the current rounded percentages
+  const afterRoundSum = rounded.reduce((pre, curr) => pre + curr, 0)
+
+  // get the amount of numbers larger or equal to 1
+  const countMutableItems = rounded.filter((x) => x >= 1).length
+
+  // get the current error rate based off of the expected total and calculated total
+  const errorRate = 100 - afterRoundSum
+
+  // calculate the amount needed to subtract from the rounded percentages
+  const deductPortion = Math.ceil(errorRate / countMutableItems)
+
+  // sort the rounded percentages in desc order
+  const biggest = [...rounded]
+    .sort((a, b) => b - a)
```

Comment:
This sorts them in descending order based on the size of the number, but shouldn't it be descending order on the remaining percentage?

Starting with these percentages: 1.7, 2.3, 96.0, 0
ends up rounding to 1, 2, 97, 0